### PR TITLE
fix(ras-acc): replace email param with secret for magic links

### PR DESCRIPTION
**Pull request by chickenn00dle**
### All Submissions:

* [ ] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/trunk/.github/CONTRIBUTING.md)?
* [ ] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Closes https://app.asana.com/0/1207817176293825/1208379852549384/f

This PR removes the email param from magic links and instead uses a random secret which is stored as user meta when the magic link is generated.

Screenshot 2024-10-14 at 16 25 38

Note: With these changes the login link will no longer be good after logging in via the auth flow. I'm not sure if this is expected, but I imagine these links should not work permanently. Let me know if this is not the case and the secret should be stored long term.

### How to test the changes in this Pull Request:

- As a logged out reader, attempt to login via the auth modal
- In the OTP email, verify the magic link no longer contains an email address as a param
- Click the magic link and verify you are logged in
- Log out
- Click the magic link in the same email once again and verify you are not able to be authenticated this time

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/class-magic-link.php [hunk 1]
```diff
@@ -379,6 +407,19 @@ public static function generate_token( $user ) {
...
 
+	/**
+	 * Generate secret.
+	 *
+	 * @param \WP_User $user User to generate the secret for.
+	 *
+	 * @return string
+	 */
+	public static function generate_secret( $user ) {
+		$secret = str_replace( '-', '', wp_generate_uuid4() );
```
Comment (chickenn00dle) [inline includes/class-magic-link.php]: Seems its possible for `wp_generate_uuid4` to generate duplicates in some rare instances, but given we are only storing these secrets between magic link creation and login and at most until logout, I think it should be fine.

But let me know if you think we should use something else here!

File: includes/class-magic-link.php [hunk 2]
```diff
@@ -1153,7 +1197,7 @@ public static function newsletters_email_verification_email( $email, $user, $url
 		$verification_url   = \add_query_arg(
 			[
 				'action'   => self::AUTH_ACTION,
-				'email'    => urlencode( $user->user_email ),
+				'email'    => self::generate_secret( $user ),
```
Comment (chickenn00dle) [inline includes/class-magic-link.php]: I'm not sure how to trigger `newsletters_email_verification_email` to test in the UI. Would love some direction on this.

File: includes/class-magic-link.php [hunk 3]
```diff
@@ -729,39 +772,41 @@ function () {
...
-			$email = \sanitize_email( $_GET['email'] );
-			if ( $email ) {
-				$user = \get_user_by( 'email', $email );
-				if ( ! $user ) {
+			$secret = filter_input( INPUT_GET, 'secret', FILTER_SANITIZE_STRING );
+			if ( $secret ) {
+				$users = \get_users(
+					[
+						'meta_key'   => self::USER_SECRET_META,
+						'meta_value' => $secret, // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_value
```
Comment (chickenn00dle) [inline includes/class-magic-link.php]: Not sure if there is a more performant way to handle this. Let me know!

File: includes/class-magic-link.php (hunk 2 above)
Comment (miguelpeixe) [reply includes/class-magic-link.php]: This is Newspack Newletters' email verification for managing newsletters subscriptions. Under RAS this verification is automatically done when the reader sets a password or connects via magic link/otp.

You can test it by manually removing the newsletter verification in wp shell: `wp user meta delete {user_id} newspack_newsletters_email_verified`. Visit "My Account -> Newsletters" and click to verify.

File: includes/class-magic-link.php (hunk 1 above)
Comment (miguelpeixe) [reply includes/class-magic-link.php]: With a rotating secret, I also think this is more than enough entropy.

File: includes/class-magic-link.php (hunk 2 above)
Comment (chickenn00dle) [reply includes/class-magic-link.php]: Thanks! Just tested and seems to be working. Also noting I added a quick small commit to fix updating the key here: https://github.com/Automattic/newspack-plugin/pull/3476/commits/6e813075b6c7d6e608fa15f76cab0ce5c85bc539

**Comment by miguelpeixe** [comment]
With the proposed approach only the last magic link is going to work because the user meta value is replaced every token request. We currently support [5 tokens](https://github.com/Automattic/newspack-plugin/blob/bfe0b8769856f26c57cb82dbed058e1474229e39/includes/class-magic-link.php#L353-L354) and the reader should be able to log in from any of the last 5.

Given this is for obfuscating the reader's email from the link, there isn't a security feature by rotating it, so I suggest having a single permanent token created once. This increases our chance of collision with uuid, so we can use `wp_hash( $user_email )` instead, which ensures uniqueness and is safely salted.

**Comment by miguelpeixe** [comment]
> We currently support [5 tokens](https://github.com/Automattic/newspack-plugin/blob/bfe0b8769856f26c57cb82dbed058e1474229e39/includes/class-magic-link.php#L353-L354) and the reader should be able to log in from any of the last 5.

We should add a unit test for that 😄

**Comment by chickenn00dle** [comment]
Thanks for the review @miguelpeixe!

> Given this is for obfuscating the reader's email from the link, there isn't a security feature by rotating it, so I suggest having a single permanent token created once. This increases our chance of collision with uuid, so we can use wp_hash( $user_email ) instead, which ensures uniqueness and is safely salted.

Alright. That makes sense. I've updated to persist the secret in https://github.com/Automattic/newspack-plugin/pull/3476/commits/8bdcb1b52650a6340fcb508c064715f8cd95e3d7

> We should add a unit test for that 😄

I did not add a unit test for this specific case, but did get one added for the `generate_secret` method added in this PR: https://github.com/Automattic/newspack-plugin/pull/3476/commits/129d8d4eca8396f41e899ace26e92c256ec4368d

**Comment by chickenn00dle** [comment]
> We should add a unit test for that 😄

Nevermind, I went ahead and added a unit test for the 5+ tokens case: https://github.com/Automattic/newspack-plugin/pull/3476/commits/7f505def12990af323850ac768e58fb2f116cce7

File: includes/class-magic-link.php [hunk 4]
```diff
@@ -729,39 +754,40 @@ function () {
...
+					]
+				);
+				if ( empty( $users ) || count( $users ) > 1 ) {
```
Comment (miguelpeixe) [inline includes/class-magic-link.php]: Non-blocking performance tweak:

```suggestion
				$user_query = new \WP_User_Query(
					[
						'meta_key'    => self::USER_SECRET_META,
						'meta_value'  => $secret, // phpcs:ignore WordPress.DB.SlowDBQuery.slow_db_query_meta_value
						'number'      => 1,
						'count_total' => false,
					]
				);
				$users = $user_query->get_results();
				if ( empty( $users ) ) {
```

**Review by miguelpeixe — APPROVED** [review]
Thank you for adding unit tests! It's working great. Left a non-blocking suggestion for performance.

File: includes/class-magic-link.php (hunk 4 above)
Comment (chickenn00dle) [reply includes/class-magic-link.php]: Thanks for the suggestion! Updated in https://github.com/Automattic/newspack-plugin/pull/3476/commits/81af75c89b982391bd23766d67ede80822d4ea10

**Pull request merged**